Add typing information to external API

leonhard-s opened the pull request:
As mentioned in #104, I have had a typed version locally for a while to satisfy my own linters. However my version was using the Python 3.5 type annotation syntax and utilities, which would break the Python 2.7 compatibility of the package. So, I converted them to the Python 2 type comment syntax for this PR.

Not exactly thrilled about those `try`-block guarded imports, but it is the only way to properly type Python 2 compatible code. Python 2.7 may be end-of-life, but this does not feel worth breaking compatibility over.

I also only covered the external API, i.e. any methods that are included in `__init__.py`'s `__all__` list. None of the runtime code has been touched, but I also have not validated this draft on versions other than Python 3.8.5 yet.

bgreen-litl commented:
Thanks for this.  I like how you did the python 2.7 support and I don't have any problem with the try block imports.  That said, python 2.7 is EOL now, and I wonder if we what we should really be doing is releasing a backoff 2.0 which drops 2.7 support and uses python 3 hinting syntax.

leonhard-s commented:
I added [another branch](https://github.com/leonhard-s/backoff/commit/16f4e02a005f96f8ffa3d9ee0569f6bf22b388ca) to my fork that uses the Python 3 type annotations instead. I think it's fine to drop Python 2.7 in the next major update, I just didn't want to force that incompatibility just for the type hints.

Are there other aspects you would revisit for a backoff 2.0 or mostly just discarding the Python 2.7 code?

michael0liver commented:
+1 for dropping Python 2!

hoefling commented:
@bgreen-litl gentle bump. What's your current view on the type hints problematic? We also have type stubs for `backoff` and would gladly move them in the upstream. I think dropping Python 2.7 (maybe after releasing a 1.11 from the current master as the last one supporting 2.7) and introducing type hints would be great news to `backoff` users (at least for me :blush:)

bgreen-litl commented:
I do want to get this done. I think your suggestion of doing one more 2.7 supporting release with the current master is a good one.

hoefling commented:
@bgreen-litl I should've look a bit more before writing the comment :blush: thought this wasn't developed anymore since Oct 2020, until I've discovered the `python-3` branch. Sorry for the noise!

bgreen-litl commented:
@hoefling @michael0liver @leonhard-s i just pushed API type hints along with dropping 2.7 support to master.  Please feel free to review/comment if you are so inclined.  Plan is to release this as backoff 2.0

bgreen-litl commented:
@hoefling oh, also i released backoff 1.11.0 which was master with python 2.7 deprecated

leonhard-s commented:
Thanks for the update!

I had to make my changes on another branch as GitHub won't allow me to revive this one: #127

These are the changes:

- Added the `py.typed` marker file (some type checkers will not parse type hints without them)
- Flagged optional arguments as explicitly optional (making the default value "None" is no longer enough according to PEP484)
- Updated decorator return types to avoid hiding the original function's signature

hoefling commented:
@bgreen-litl the type hints are awesome and replace our stubs with no errors, great stuff! However, IMO the changes proposed in #127 should be included in the 2.0 release, otherwise `mypy` won't recognize the inline type hints without workarounds.

@leonhard-s `TypedDict` could be only introduced for the `_Handler` first, smth like
```py
if sys.version_info >= (3, 8):
    from typing import TypedDict
    class _Details(TypedDict, total=False):
        target: Any
        ...
else:
    _Details = Dict[str, Any]

_Handler = Callable[[_Details], None]
```

leonhard-s commented:
@hoefling That is what I was doing (though I didn't know about disabling totality, neat), but the user would still have to either provide their own compatible `TypedDict` or be given access to backoff's `_Details` type to be able type their handler functions in a compatible manner if I'm not mistaken.

michael0liver commented:
I think that `_Details` type should be exposed as public API for end users to reuse it. Also for users below Python 3.8 we can use the official backport from [`typing_extensions`](https://github.com/python/typing/tree/master/typing_extensions) package:
```python
import sys

if sys.version_info >= (3, 8):
    from typing import TypedDict
else:
    from typing_extensions import TypedDict


class Details(TypedDict, total=False):
    ...

```

hoefling commented:
@leonhard-s ah, you're right - `Details` should be exposed then, and it's something for @bgreen-litl to decide.
@michael0liver good idea, though this would make `typing_extensions` a runtime dependency. We could go with a slightly mode complex construct here, but now I'm unsure this is worth the fuzz:
```py
from typing import TYPE_CHECKING

if sys.version_info >= (3, 8):
    from typing import TypedDict
elif TYPE_CHECKING:
    from typing_extensions import TypedDict
else:
    Details = Dict[str, Any]
```

bgreen-litl commented:
> @bgreen-litl the type hints are awesome and replace our stubs with no errors, great stuff! However, IMO the changes proposed in #127 should be included in the 2.0 release, otherwise `mypy` won't recognize the inline type hints without workarounds.

Yup, that was good PR. It is now merged to master and will be part of 2.0

> ```python
>     class _Details(TypedDict, total=False):
>         target: Any
>         ...
> ```

As a side effect of this conversation, you got me curious about the above syntax. I understand what `total=False` is doing, but where does this syntax come from?  I don't think I knew about class declarations having valued arguments like this, apparently after the list of inherited classes.  What else is this used for?

leonhard-s commented:
> As a side effect of this conversation, you got me curious about the above syntax. [...] where does this syntax come from?

In Python 3.9, `TypedDict` is apparently a function to allow non-class-based syntaxes like `Point2D = TypedDict('Point2D', x=int, y=int, label=str)` ([`TypedDict` source](https://github.com/python/cpython/blob/05162993fe62e7fa3acebdd0062586b9bf63d46a/Lib/typing.py#L2269)). For subclasses there is the [`__init_subclass__`](https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__) hook on the base class.

What are your thoughts on making the `_Details` type visible to the user and restricting the `_Handler` type to require it? It does add a few unneeded names into the runtime namespace but also makes the keys and types in the backoff details dictionary visible.

bgreen-litl commented:
> > As a side effect of this conversation, you got me curious about the above syntax. [...] where does this syntax come from?
> 
> In Python 3.9, `TypedDict` is apparently a function to allow non-class-based syntaxes like `Point2D = TypedDict('Point2D', x=int, y=int, label=str)` ([`TypedDict` source](https://github.com/python/cpython/blob/05162993fe62e7fa3acebdd0062586b9bf63d46a/Lib/typing.py#L2269)).
> 
> For classes, there is the __init_subclass__ hook on the base class that can add some metaclass-like customisation of subclasses.

Thanks, I didn't know about this keyword arg customization of subclasses.

> What are your thoughts on making the `_Details` type visible to the user and restricting the `_Handler` type to require it? It does add a few unneeded names into the runtime namespace but also makes the keys and types in the backoff details dictionary visible.

I'm trying to work through what I think of it.  Based on the snippets you guys worked out above, I came up with this so far:

```python
if sys.version_info >= (3, 8):
    from typing import TypedDict
elif TYPE_CHECKING:
    from typing_extensions import TypedDict

try:
   TypedDict
except NameError:
    Details = Dict[str, Any]
else:
    class Details(TypedDict, total=False):
        """A nice docstring here"""
        target: Callable[..., Any]
        args: List[Any]
        kwargs: Dict[str, Any]
        tries: int
        elapsed: float
        wait: float  # on_backoff handler only
        value: Any  # on_predicate decorator only
```

And then we'd expose it publicly as `backoff.Details` I suppose.  I have historically been fairly stingy about adding anything to the public API, but I think there's probably a pretty good argument for this.  There have also intentionally never been any runtime dependencies for `backoff`, but if `typing_extensions` is optional (and official), I can probably convince myself it's a valid exception.

I think we could also split Details into separate types depending on the decorator/handler and drop `total=False` but that strikes me as pretty overkill.

leonhard-s commented:
We could move `Details` into a separate namespace like `backoff.types.Details`, without reimporting it to `__init__.py`.
That way users wanting to use static type hints can import it via `from backoff.types import Details`, but it would be independent from the main API and would not be accessible via `backoff.Details`.

Regarding `typing_extensions`, we could just use it if available, and silently fall back to `Dict[str, Any]` if it isn't. That way there are no runtime or development dependencies, but users can still harness the more specific `TypedDict` hint if they have it installed:
```py
import sys
from typing import TYPE_CHECKING

if sys.version_info >= (3, 8):
    from typing import TypedDict
elif TYPE_CHECKING:
    try:
        from typing_extensions import TypedDict
    except ImportError:
        pass
try:
    _ = TypedDict
except NameError:
    ... # Same as before
```

If we do decide to expose the `Details` type, I think it makes sense to split it out into its decorator-specific variants. The main reason to type it at all is to have certainty of which keys are available, those keys missing at runtime might be more confusing than not having a typed dictionary at all.

bgreen-litl commented:
Just want to indicate for @hoefling and @michael0liver that this conversation has migrated to over #128 if either want to comment there.

bgreen-litl commented:
Type hints are in master now.  Thanks for this!

michael0liver commented:
Great work.
